crates/net/downloaders/src/bodies/bodies.rs

```diff
@@ -205,28 +212,50 @@ where
         self.download_range = RangeInclusive::new(1, 0);
         self.latest_queued_block_number.take();
         self.in_progress_queue.clear();
-        self.buffered_responses.clear();
         self.queued_bodies.clear();
+        self.buffered_responses.clear();
+        self.num_buffered_blocks = 0;
 
+        // reset metrics
         self.metrics.in_flight_requests.set(0.);
         self.metrics.buffered_responses.set(0.);
+        self.metrics.buffered_blocks.set(0.);
+        self.metrics.queued_blocks.set(0.);
     }
 
     /// Queues bodies and sets the latest queued block number
     fn queue_bodies(&mut self, bodies: Vec<BlockResponse>) {
         self.latest_queued_block_number = Some(bodies.last().expect("is not empty").block_number());
         self.queued_bodies.extend(bodies.into_iter());
+        self.metrics.queued_blocks.set(self.queued_bodies.len() as f64);
+    }
+
+    /// Removes the next response from the buffer.
+    fn pop_buffered_response(&mut self) -> Option<OrderedBodiesResponse> {
+        let resp = self.buffered_responses.pop()?;
+        self.metrics.buffered_responses.decrement(1.);
+        self.num_buffered_blocks -= resp.0.len();
+        self.metrics.buffered_blocks.set(self.num_buffered_blocks as f64);
+        Some(resp)
+    }
+
+    /// Adds a new response to the internal buffer
+    fn buffer_bodies_response(&mut self, response: Vec<BlockResponse>) {
+        self.num_buffered_blocks += response.len();
+        self.metrics.buffered_blocks.set(self.num_buffered_blocks as f64);
+        let response = OrderedBodiesResponse(response);
+        self.buffered_responses.push(response);
+        self.metrics.buffered_responses.set(self.buffered_responses.len() as f64);
```

Comment:
nice, this is a good way to make sure we keep track of number of blocks instead of number of requests